packages/next-admin/src/types.ts

```diff
@@ -341,6 +342,70 @@ export type ListOptions<T extends ModelName> = {
   filters?: FilterWrapper<T>[];
 };
 
+export type SubmitResourceResponse =
+  | {
+      error: string;
+      updated?: undefined;
+      data?: undefined;
+      redirect?: undefined;
+      created?: undefined;
+      createdId?: undefined;
+      validation?: undefined;
+    }
+  | {
+      error: string;
+      updated?: undefined;
+      data?: undefined;
+      redirect?: undefined;
+      created?: undefined;
+      createdId?: undefined;
+      validation: PropertyValidationError[];
+    }
+  | {
+      error?: undefined;
+      updated: true;
+      data: any;
+      redirect: boolean;
+      created?: undefined;
+      createdId?: undefined;
+      validation?: undefined;
+    }
+  | {
+      error?: undefined;
+      updated?: undefined;
+      data: any;
+      redirect: boolean;
+      created: true;
+      createdId: any;
+      validation?: undefined;
+    };
+
+export type EditModelHooks = {
+  /**
+   * a function that is called before the form data is sent to the database.
+   *
+   * @param data - the form data as a record
+   * @param mode - the mode of the form, either "create" or "edit"
+   * @param request - the request object. Note: on App Router, it will not be possible to call `formData` as the handler already calls it once
+   * @returns the transformed form data record or a Response or NextApiResponse.
+   * @throws HookError - if the hook fails, the status and message will be sent in the handler's response
+   */
+  beforeDb?: (
+    data: Record<string, string | UploadParameters | null>,
+    mode: "create" | "edit",
+    request: NextRequest | NextApiRequest
+  ) => Promise<Record<string, string | UploadParameters | null>>;
+  /**
+   * a function that is called after the form submission. It takes the response of the db insertion as a parameter.
+   * @param data
+   */
+  afterDb?: (
+    data: SubmitResourceResponse,
+    mode: "create" | "edit",
+    request: NextRequest | NextApiRequest
+  ) => Promise<void>;
```

Comment:
Perhaps this function should return the data object to be identical to `beforeDb` return type.

Reply:
This will actually return the same type as the `data` argument, in which the data response can be modified